querqy-for-lucene/querqy-lucene/src/main/java/querqy/lucene/rewrite/MultiMatchDismaxQueryStructurePostProcessor.java

```diff
@@ -85,33 +123,198 @@ protected void processTopLevelDmq(final DisjunctionMaxQueryFactory dmq) {
         final int numDisjuncts = dmq.getNumberOfDisjuncts();
         if (numDisjuncts > 1) {
 
-            final Map<Object, List<LuceneQueryFactory<?>>> grouped = dmq.disjuncts.stream()
-                    .collect(Collectors.groupingBy(this::getGroupingValue));
+            final Map<String, List<LuceneQueryFactory<?>>> factoriesByField = RegroupDisjunctsByFieldProcessor
+                    .regroupByFields(dmq.disjuncts);
 
-            grouped.values().stream()
-                    .filter(factories -> (numDisjuncts > factories.size()) && factories.size() > 1)
-                    .forEach(factories -> applyDisjunctGrouping(dmq, factories));
+            dmq.disjuncts.clear();
 
-            if (numDisjuncts > dmq.getNumberOfDisjuncts()) {
-                dmq.setTieBreaker(multiMatchTieBreakerMultiplier);
+            for (final Map.Entry<String, List<LuceneQueryFactory<?>>> entry : factoriesByField.entrySet()) {
+                final List<LuceneQueryFactory<?>> factories = entry.getValue();
+                if (factories.size() == 1) {
+                    dmq.add(factories.get(0));
+                } else {
+                    dmq.add(new DisjunctionMaxQueryFactory(factories, multiMatchTieBreakerMultiplier));
+                }
             }
+            dmq.setTieBreaker(dmqTieBreakerMultiplier);
 
         }
     }
 
-    protected Object getGroupingValue(final LuceneQueryFactory<?> factory) {
-        if (factory instanceof TermSubQueryFactory) {
-            return ((TermSubQueryFactory) factory).sourceTerm.getValue();
-        } else if (factory instanceof TermQueryFactory) {
-            return ((TermQueryFactory) factory).sourceTerm.getValue();
-        } else return factory;
+    public static class RegroupDisjunctsByFieldProcessor extends LuceneQueryFactoryVisitor<Void> {
```

Comment:
This would require us to change the visibility of some factory properties from package to public, which would need more careful consideration. Leaving it as inner classes for now.